command/transfer_coin.go

```diff
@@ -14,64 +18,140 @@ import (
 type TransferCoinCommand struct{}
 
 func (c *TransferCoinCommand) Run(msg string, event *CoinEvent) (string, error) {
-	re := regexp.MustCompile(`^(?i)transfer (?P<amount>[0-9]+) (?:to )?\<@(?P<to_slack_id>[0-9A-Z]+)\> (?:for)??(?P<reason>.+)`)
-	matches := re.FindStringSubmatch(event.Message)
-
-	if len(matches) < 4 {
-		return "Invalid transfer format. Expected something like `@Alex Koin transfer 3 to @alexk for being amazing`. See the channel details for command syntax.", nil
+	parsedResult, err := parseMessage(event.Message)
+	if err != nil {
+		return err.Error(), nil
 	}
 
-	amount, err := strconv.Atoi(matches[1])
-	if err != nil {
-		log.Printf("INF amount not parsed as int: %s", matches[1])
-		return "Invalid transfer amount.", nil
+	// parsing error should be covered in func validateMessage
+	totalAmount, _ := strconv.Atoi(parsedResult["amount"])
+	toUserIds := strings.Split(parsedResult["to_slack_ids"], ",")
+	reason := parsedResult["reason"]
+	splitAmounts := splitCoins(totalAmount, len(toUserIds))
+
+	for i, toUserId := range toUserIds {
```

Comment:
Is there a way to wrap this in a transaction? I'm just worried that if someone spams a few transfers and these are going one at a time, that we end up in a bad state.

Reply:
good point, I actually tried to move all the validation part to the beginning of the method, to reduce the possibility of partial transfer, but might be worth to just use a transaction wrapper, will take a look of the support of GORM